tests/dataset_runner.py

```diff
@@ -369,20 +369,35 @@ def assert_data_browser_bundles(self):
         ).to_return_value(value=True)
 
     def _assert_data_browser_bundles(self, project_shortname):
+        # Numbers here correspond to references to files for each bundle (AKA contributions). We must wait for all files
+        # references to be indexed before we continue to cleanup in order to avoid potential races.
+        # A file is ref is a contribution if it appears in the bundle with "indexed": false.
+        # For ss2 we have 2 fastqs in the primary bundle; these 2 and 18 other contributions from the secondary
+        # For optimus we have 3 fastqs in the primary bundle; these 3 and 7 other contributions from the secondary
+        # for a total of 22 and 13 contributions respectively
+        num_contributions_by_project = {'Smart-seq2': 22, 'optimus': 13}
+        expected_bundle_uuids = set(self.primary_bundle_uuids).union(self.secondary_bundle_uuids)
+        files = self.azul_agent.get_entities_by_project('files', project_shortname)
+        bundle_uuids = {bundle['bundleUuid'] for file in files for bundle in file['bundles']}
+        project_shortnames = {project_short_name
+                              for file in files
+                              for project in file['projects']
+                              for project_short_name in project['projectShortname']}
+        num_contributions = sum(1 for f in files for b in f['bundles'])
         try:
-            expected_bundle_uuids = set(self.primary_bundle_uuids).union(self.secondary_bundle_uuids)
-            files = self.azul_agent.get_entities_by_project('files', project_shortname)
-            bundle_uuids = {bundle['bundleUuid'] for file in files for bundle in file['bundles']}
             assert bundle_uuids == expected_bundle_uuids
-            project_shortnames = {project_short_name for file in files for project in file['projects']
-                                  for project_short_name in project['projectShortname']}
             assert project_shortnames == {project_shortname}
+            assert num_contributions == num_contributions_by_project[self.dataset.name]
```

Comment:
If `num_contributions < num_contributions_by_project[self.dataset.name]` changes to `num_contributions > num_contributions_by_project[self.dataset.name]`, this would time out even though we know the `>` case should never occur and constitutes an error. I think you can simplify this by not doing the `try: assert except AssertionError` which smells funny anyways. Simply `if … < …: return False`; elif … == …: return True; else: assert False` or equivalent. The `bundle_uuids` and `project_shortnames` asserts would have to be adjusted similarly.

Reply:
It would not time out because I added another assert in the `finally` clause https://github.com/HumanCellAtlas/dcp/pull/326/files#diff-b4fb3043d409f4155bdd0e6697aeaf78R400 I can change the design though.